### ISSUE TITLE
 1.1.8 not work when using multi-root workspace

### ISSUE THREAD

**Issue by NateScarlet** [microsoft/pyright#395]
**Describe the bug**
After i upgraded to 1.1.8, i can not get any intellisense when using multi-root workspace.
It works if i switch to single root workspace for same folder.

**To Reproduce**
Upgrade to 1.1.8 and use multi-root workspace.

**Expected behavior**
Work as 1.1.7

**Screenshots or Code**
pyright 1.1.7 output

```
Pyright language server starting
Starting service instance "web"
Starting service instance "web-doc"
Starting service instance "server"
Starting service instance "server-doc"
Starting service instance "{project}"
Fetching settings for workspace(s)
Received updated settings
Searching for source files
No source files found.
No configuration file found.
Setting pythonPath for service "web": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
No source files found.
No configuration file found.
Setting pythonPath for service "web-doc": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
No source files found.
Loading configuration file at d:\Users\{username}\src\{path}\server\pyrightconfig.json
No include entries specified; assuming d:\Users\{username}\src\{path}\server\pyrightconfig.json
Setting pythonPath for service "server": "d:\Users\{username}\src\{path}\server\.venv\Scripts\python.exe"
typingsPath d:\Users\{username}\src\{path}\server\typings is not a valid directory.
Searching for source files
Found 300 source files
No configuration file found.
Setting pythonPath for service "server-doc": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
No source files found.
No configuration file found.
Setting pythonPath for service "{project}": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
No source files found.
```

pyright 1.1.8 output

```
Pyright language server starting
Starting service instance "web"
Starting service instance "web-doc"
Starting service instance "server"
Starting service instance "server-doc"
Starting service instance "{project}"
Fetching settings for workspace(s)
Received updated settings
No configuration file found.
Setting pythonPath for service "web": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
```

**VS Code extension or command-line**
VSCode extention: 1.1.8

**Additional context**
`web` is using npm, maybe this caused by too many files in `node_modules` ?

**Comment by erictraut** [comment]
Thanks for reporting the problem. 

This is undoubtedly caused by this change: "Changed the behavior of VS Code extension when no pyrightconfig.json is present. It now defaults to include all source files under the root workspace directory. This was already the behavior for the command-line version of pyright."

The last line of the 1.1.8 output ("Searching for source files") should eventually be followed by "Found X source files". If you let it run for a while, does it eventually complete? I suspect that you have a lot of files, and it's just taking a long time to scan for all the python files in your workspace. Does it appear that pyright is consuming a bunch of CPU and disk I/O (e.g. if you look in the Activity Monitor or Task Manager)?

I recommend adding this to your pyrightconfig.json file(s):
```
{
  "exclude": [
    "**/node_modules",
    "**/__pycache__"
  ]
}
```

I can't yet explain why switching from a multi-root workspace to a single-root workspace would change this behavior.

**Comment by NateScarlet** [comment]
not work even i added pyrightconfig.json to `web`

multi-root output
```
Pyright language server starting
Starting service instance "web"
Starting service instance "web-doc"
Starting service instance "server"
Starting service instance "server-doc"
Starting service instance "{project}"
Fetching settings for workspace(s)
Received updated settings
Searching for source files
No source files found.
Loading configuration file at d:\Users\{username}\src\{path}\web\pyrightconfig.json
No include entries specified; assuming d:\Users\{username}\src\{path}\web\pyrightconfig.json
Setting pythonPath for service "web": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
typingsPath d:\Users\{username}\src\{path}\web\typings is not a valid directory.
Searching for source files
No source files found.
No configuration file found.
Setting pythonPath for service "web-doc": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
Found 1 source file
Loading configuration file at d:\Users\{username}\src\{path}\server\pyrightconfig.json
Setting pythonPath for service "server": "d:\Users\{username}\src\{path}\server\.venv\Scripts\python.exe"
typingsPath d:\Users\{username}\src\{path}\server\typings is not a valid directory.
Searching for source files
```

single-root output
```
Pyright language server starting
Starting service instance "server"
Fetching settings for workspace(s)
Received updated settings
Loading configuration file at d:\Users\{username}\src\{path}\server\pyrightconfig.json
Setting pythonPath for service "server": "d:\Users\{username}\src\{path}\server\.venv\Scripts\python.exe"
typingsPath d:\Users\{username}\src\{path}\server\typings is not a valid directory.
Searching for source files
Found 298 source files
```

**Comment by NateScarlet** [comment]
![image](https://user-images.githubusercontent.com/21083014/70296093-4d1f2f00-1824-11ea-963d-73032bd6d474.png)

and a Code.exe with argument contains `ms-pyright.pyright-1.1.8\server\server.bundle.js` keep using 13% CPU (8 core), memory usage keep increasing.
i will let it run for while to see whether it can complete.

**Comment by NateScarlet** [comment]
It completed in 30 minutes or less.
```
Pyright language server starting
Starting service instance "web"
Starting service instance "web-doc"
Starting service instance "server"
Starting service instance "server-doc"
Starting service instance "{project}"
Fetching settings for workspace(s)
Received updated settings
Searching for source files
No source files found.
Loading configuration file at d:\Users\{username}\src\{path}\web\pyrightconfig.json
No include entries specified; assuming d:\Users\{username}\src\{path}\web\pyrightconfig.json
Setting pythonPath for service "web": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
typingsPath d:\Users\{username}\src\{path}\web\typings is not a valid directory.
Searching for source files
No source files found.
No configuration file found.
Setting pythonPath for service "web-doc": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
Found 1 source file
Loading configuration file at d:\Users\{username}\src\{path}\server\pyrightconfig.json
Setting pythonPath for service "server": "d:\Users\{username}\src\{path}\server\.venv\Scripts\python.exe"
typingsPath d:\Users\{username}\src\{path}\server\typings is not a valid directory.
Searching for source files
Found 298 source files
No configuration file found.
Setting pythonPath for service "server-doc": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
Found 3583 source files
No configuration file found.
Setting pythonPath for service "{project}": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
Searching for source files
Found 6966 source files
``

{project} is the parent folder of other 4 folder.

**Comment by NateScarlet** [comment]
it fixed if i add pyrightconfig.json to {project} folder
```json
{
  "exclude": ["**"]
}
```

seems some kind of loop caused by folder stucture like
```
- {project}
| - {python-code}
```
and add both {project} and {python-code} to same workspace.

**Comment by NateScarlet** [comment]
it work if i copy all files exclude `node_modules` to `web-with-no_node_modules`

```
Pyright language server starting
Starting service instance "server"
Starting service instance "web-with-no-node_modules"
Fetching settings for workspace(s)
Received updated settings
Loading configuration file at d:\Users\{username}\src\{path}\server\pyrightconfig.json
Setting pythonPath for service "server": "d:\Users\{username}\src\{path}\server\.venv\Scripts\python.exe"
typingsPath d:\Users\{username}\src\{path}\server\typings is not a valid directory.
Searching for source files
Found 298 source files
Loading configuration file at d:\Users\{username}\src\{path}\web-with-no-node_modules\pyrightconfig.json
No include entries specified; assuming d:\Users\{username}\src\{path}\web-with-no-node_modules\pyrightconfig.json
Setting pythonPath for service "web-with-no-node_modules": "C:\Users\{username}\AppData\Local\Programs\Python\Python37\python.exe"
typingsPath d:\Users\{username}\src\{path}\web-with-no-node_modules\typings is not a valid directory.
Searching for source files
No source files found.
```

and not work if i only keep `server` and `web` folder in workspace.

seems `exclude` in `web/pyrightconfig.json` has no affect.

**Comment by erictraut** [comment]
OK, thanks for providing that additional information. 

Within your workspace definition file ("X.code-workspace"), there should be a "folders" section. I suspect that one of the folders is ".", which corresponds to your "{project}" directory. Unless you have a good reason for including this as a workspace root, I recommend that you remove it.

The problem is that pyright is scanning multiple large file hierarchies looking for ".py" and ".pyi" files. In your case.

I'm going to make a few changes to pyright that will help mitigate this problem. If there is no pyrightconfig.json file provided for a workspace root, I'm going to assume that the following directories should be excluded from the scan: "**/node_modules" and "**/__pycache__".

**Comment by erictraut** [comment]
This is now implemented in version 1.1.9, which I just published.

**Comment by NateScarlet** [comment]
I tried 1.1.9, it takes more than 2 minitues to found all 305 source files.
In 1.1.7 or single-root workspace, it complete immediately.

And i can't reproduce the problem with other folder that `node_modules` has similar size.

**Comment by NateScarlet** [comment]
even i use a multi-root workspace that only have this one root,
it still slower than single-root workspace

**Comment by erictraut** [comment]
Can you provide the "code-workspace" file that defines your workspace? Does it contain a folder definition for the root directory (".")? If so, what happens if you remove that?

**Comment by NateScarlet** [comment]
original workspace
```
  "folders": [
    {
      "name": "web",
      "path": "web"
    },
    {
      "name": "server",
      "path": "server"
    },
    {
      "name": "web-doc",
      "path": "web-doc"
    },
    {
      "name": "server-doc",
      "path": "server-doc"
    },
    {
      "name": "all",
      "path": "."
    }
  ],
```

Not work at 1.1.8, but work at 1.1.9 :
```
    {
      "name": "web",
      "path": "web"
    },
    {
      "name": "server",
      "path": "server"
    },
```

slow at 1.1.9, even i have pyrightconfig.json `exclude: ["*"]` in `server-doc`
```
  "folders": [
    {
      "name": "server",
      "path": "server"
    },
    {
      "name": "server-doc",
      "path": "server-doc"
    }
  ],
```
`server-doc` has 16k files (include python venv for sphinx)

**Comment by erictraut** [comment]
An exclude of `"*"` will only exclude files at the top-level directory. I think you want to exclude your venv and all files within it (`"venv"` - or whatever that directory is named).

Out of curiosity, what is your venv file named? Is it "venv"? Maybe I should add that name to the default exclusion list?

Prior to 1.1.7, pyright didn't scan for python files if there was no pyrightconfig.json file or no include section in the pyrightconfig.json file. The problem is that some features (like "rename symbols" or "find references") require pyright to know about which python source files are considered part of your program. To make these features work, pyright needs to scan for python files in each workspace. If you have a large number of files, this will take a while. You can use the exclude section of pyrightconfig.json to prevent pyright from scanning certain subfolders.

**Comment by NateScarlet** [comment]
does file get included if parent folder already excluded?
venv folder name is `.venv`, and `.git` folder should also be excluded.

**Comment by erictraut** [comment]
No, if the parent folder is excluded, its children will also be excluded.

I guess that ".venv" is the common name for virtual environments that are located within workspace. (I always store my virtual environments outside of the workspace, but I know that many users like to place it within the workspace.) I should probably exclude ".venv" by default. The same goes for ".git".

Thanks for the suggestions.

**Comment by NateScarlet** [comment]
I got same search speed after added this pyrightconfig.json to all other folders in workspace:

```json
{
  "include": ["NOTHING"]
}
```

seems exclude all from include will still cost extra resources.

**Comment by erictraut** [comment]
@MikhailArkhipov & @jakebailey do you have any advice on this one? The best idea I've come up with is to exclude any `.venv` and `.git` directory by default if there no exclude paths are defined in the config file. I already exclude `**/node_modules` and `**/__pycache__` by default, so this approach would simply expand that list.

**Comment by jakebailey** [comment]
MPLS is open file driven. The only action we perform on the entire workspace is symbol indexing for the workspace symbol search (and for offering import suggestions). This is performed independently. We take the excludes defined by the client (typically the VS Code defaults like `**/node_modules` + some python specific ones set by the extension), then use that to skip over files in the workspace to not index. We haven't had trouble with workspace indexing for this specific purpose. You might be able to do something similar if you need to do full scanning.

In terms of references/renaming/etc, we don't use any full workspace scanning/opening to do this. Instead, we search the workspace for candidate files which contain the name being searched, then open the files which match. Then, they're analyzed as normal (as unopened user files), which means the references info is now present for querying.

As for multi-root workspace related stuff, we don't support multi-root workspaces directly; the extension spawns more than one LS per workspace (passing python versions, interpreter path, search paths, etc to each).

Hopefully I've understood the context here correctly.

**Comment by jakebailey** [comment]
Another fix for detecting folders to not look at would be to use the search paths (i.e. this is the stdlib, don't look here, this is site-packages, also skip, etc), but to do this effectively, you'd need to beef up your search path script to classify search paths like in case of pth files or something where a search path actually is user code.

**Comment by MikhailArkhipov** [comment]
Yes, we use such classification to determine what to watch for file changes (i.e. detecting `pip install`) and what is the user code.

**Comment by erictraut** [comment]
I've added ".venv" and ".git" to the default exclude paths. This will be included in the next release.

**Comment by erictraut** [comment]
The above change is implemented in 1.1.11, which I just published.
